src/iris/api.py

```diff
@@ -3311,6 +3311,32 @@ def on_post(self, req, resp):
                    WHERE `twilio_sid` = :sid''',
                 {'sid': sid, 'status': status}).rowcount
             session.commit()
+
+            if status == 'failed':
+                msg_id = session.execute(
+                    '''SELECT message_id
+                       FROM `twilio_delivery_status`
+                       WHERE `twilio_sid` = :sid''',
+                    {'sid': sid}).scalar()
+                is_retry = session.execute(
+                    '''SELECT EXISTS(SELECT * FROM `twilio_retry`
+                                     WHERE `retry_id` = :msg_id)''', {'msg_id': msg_id}).scalar()
+                # Don't retry messages that are already a retry
+                if not is_retry:
+                    retry_id = session.execute(
+                        '''INSERT INTO `message` (`created`, `incident_id`, `application_id`,
+                                                  `target_id`, `priority_id`, `body`)
+                           SELECT NOW(), `incident_id`, `application_id`, `target_id`, `priority_id`, `body`
+                           FROM `message` WHERE `id` = :msg_id
+                        ''',
+                        {'msg_id': msg_id}
+                    ).lastrowid
+                    session.execute(
+                        '''INSERT INTO `twilio_retry` (`message_id`, `retry_id`)
+                           VALUES (:msg_id, :retry_id)
+                        ''',
+                        {'msg_id': msg_id, 'retry_id': retry_id})
+            session.commit()
```

Comment:
I think this session.commit() should be inside the `if not is_retry` block as that one has the INSERTs, and we have a session.commit() earlier in the function